src/coap_blockwise.c

```diff
@@ -59,25 +64,84 @@ struct get_block_ctx
 
 /* Blockwise Uploads related functions */
 
+static void blockwise_upload_free(struct blockwise_transfer *ctx)
+{
+    if (!ctx)
+    {
+        return;
+    }
+
+    golioth_sys_sem_destroy(ctx->sem);
+    free(ctx->block_buffer);
+    free(ctx->stored_path);
+}
+
 // Function to initialize the blockwise_transfer structure for uploads
-static void blockwise_upload_init(struct blockwise_transfer *ctx,
-                                  uint8_t *data_buf,
-                                  const char *path_prefix,
-                                  const char *path,
-                                  enum golioth_content_type content_type,
-                                  read_block_cb cb,
-                                  void *callback_arg)
+struct blockwise_transfer *blockwise_upload_init(const char *path_prefix,
+                                                 const char *path,
+                                                 enum golioth_content_type content_type,
+                                                 size_t block_buffer_size,
+                                                 bool store_path)
 {
+    if ((store_path) && (strlen(path) > CONFIG_GOLIOTH_COAP_MAX_PATH_LEN))
+    {
+        GLTH_LOGE(TAG, "Path too long: %zu > %zu", strlen(path), CONFIG_GOLIOTH_COAP_MAX_PATH_LEN);
+        return NULL;
+    }
+
+    struct blockwise_transfer *ctx = golioth_sys_malloc(sizeof(struct blockwise_transfer));
+    if (!ctx)
```

Comment:
I haven't been strict about this in the past, but we should not use `!` for `NULL` comparisons. CERT recommends that the standalone `if (!x)` notation is [never used](https://wiki.sei.cmu.edu/confluence/display/c/EXP20-C.+Perform+explicit+tests+to+determine+success%2C+true+and+false%2C+and+equality). The explicit comparison is easier to read, and thus easier to maintain and less prone to bugs.
```suggestion
    if (NULL == ctx)
```